### PR TITLE
Replace string ref usage with callback refs

### PR DESCRIPTION
**Pull request by ablamunits**
Hi,

Setting element `ref`s using "String Refs" is considered [legacy API](https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs) by React and is generally discouraged due to [certain issues](https://github.com/facebook/react/pull/8333#issuecomment-271648615).

We personally ran into issues in our own setup, where the usage of string refs caused the following error to appear when `ReactTelephoneInput` was mounted in a our larger application, bundled with browserfiy:

```
removeComponentAsRefFrom(...): Only a ReactOwner can have refs. You might be removing a ref to a component that was not created inside a component's `render` method, or you have multiple copies of React loaded (details: https://fb.me/react-refs-must-have-owner).
```

I did not dig any deeper into the exact root cause of this, and perhaps this has something to do with the structure of our project. However, using callback refs in telephone input has solved the issue for us and it seems to be a reasonable solution, considering the currently used API will be deprecated at some point.

Thanks

### CODE REVIEW

File: es/ReactTelephoneInput.js [hunk 1]
```diff
@@ -173,7 +173,13 @@ export var ReactTelephoneInput = createReactClass({
         return !isEqual(nextProps, this.props) || !isEqual(nextState, this.state);
     },
     componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
-        this.setState(this._mapPropsToState(nextProps));
+        var prevFormattedNumber = this.state.formattedNumber;
+        this.setState(this._mapPropsToState(nextProps), function () {
+            // If formatted number has changed, call on change
+            if (this.state.formattedNumber !== prevFormattedNumber) {
+                this.props.onChange(this.state.formattedNumber, this.state.selectedCountry);
+            }
```
Comment (ablamunits) [inline es/ReactTelephoneInput.js]: As this component maintains its own state, if there are changes to the value of the telephone field that were caused by the incoming props these changes will be applied in `_mapPropsToState`. Prior to this fix, such a change in the state will not call `onChange`. This fixes it.

Comment (ablamunits) [reply es/ReactTelephoneInput.js]: @mukeshsoni any objections to this change? Perhaps this was suggested in the past?

Comment (mukeshsoni) [reply es/ReactTelephoneInput.js]: componentWillReceiveProps is doing a lot in this component. I shouldn't have done this in the first place. This piece of code has been the source of a lot of bugs in this component. I am thinking about completely removing componentWillReceiveProps and give an alternate solution making it a controlled/uncontrolled/(controlled + uncontrolled) component

**Comment by mukeshsoni** [comment]
@ablamunits I have completely removed the dependency on `react-dom` and also removed all the refs on all the list items. Have switched to using `react-tiny-virtual-list`, which implements windowing on the list and hence is much more performant. Win-win.

**Pull request closed without merging**